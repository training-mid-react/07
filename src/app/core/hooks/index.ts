export * from './form';
export * from './home-container';
export * from './board-container';