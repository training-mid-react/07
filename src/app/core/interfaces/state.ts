import { CHANGE_PLAYER, DROP_PIECE, RESET_GAME, SET_WINNER } from "../state/cards/actions";

export interface IGameState {
  board: number[][];
  currentPlayer: number;
  winner: number | null;
  draw: boolean;
}

export type GameAction =
  | { type: typeof DROP_PIECE; payload: number }
  | { type: typeof CHANGE_PLAYER; payload: number }
  | { type: typeof SET_WINNER; payload: number }
  | { type: typeof RESET_GAME; payload: number }

export interface IAppContextProps {
  state: IGameState;
  dispatch: React.Dispatch<any>;
}