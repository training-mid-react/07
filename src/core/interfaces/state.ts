import { GameState } from ".";

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface IState extends GameState {
}

export interface IContext {
  state: IState;
  dispatch: React.Dispatch<IAction>;
}

export interface IAction {
  type: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload: any;
}