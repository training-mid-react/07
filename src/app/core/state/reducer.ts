import { IAction, IState } from '@core/interfaces';
import { boardGameCases, boardGameInitialState } from './board-game';

export const initialState = {
    ...boardGameInitialState,
};

export const reducer = (state: IState, action: IAction): IState => {
    const cases = { ...boardGameCases };

    return cases[action.type](state, action.payload) || state;
};
