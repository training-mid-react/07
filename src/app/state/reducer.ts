import { GameState } from '../../core/interfaces';
import { IAction } from '../../core/interfaces/state';
import { boardCases, boardInitialState } from './board';

export const initialState = {
  ...boardInitialState
};

export const reducer = (state: GameState, action: IAction) => {
  const cases = { ...boardCases };
  return cases[action.type](state, action.payload) || state;
};