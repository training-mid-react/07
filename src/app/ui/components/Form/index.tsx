import { IPlayerForm } from '@core/interfaces';
import './styles.scss';
import { useCustomForm } from '@core/hooks';
import { FC } from 'react';

export const Form: FC<{
    onSubmit: (data: IPlayerForm) => void;
}> = ({ onSubmit }) => {
    const { selectOptions, register, handleSubmit, errors } = useCustomForm();

    return (
        <form className="form" onSubmit={handleSubmit(onSubmit)}>
            <h2>Configuraci√≥n de Jugadores</h2>

            <h3>Jugador 1</h3>
            <label>
                Nombre:
                <input
                    role="input"
                    type="text"
                    {...register('player1.name', {
                        required: 'Nombre es requerido',
                    })}
                    placeholder="Ingresa un nombre"
                />
                {errors.player1?.name && (
                    <span style={{ color: 'red' }}>
                        {errors.player1.name.message}
                    </span>
                )}
            </label>
            <label>
                Color:
                <select
                    {...register('player1.color', {
                        required: 'Selecciona un color',
                    })}
                >
                    {selectOptions.map((o) => (
                        <option key={o.value} value={o.value}>
                            {o.label}
                        </option>
                    ))}
                </select>
                {errors.player1?.color && (
                    <span style={{ color: 'red' }}>
                        {errors.player1.color.message}
                    </span>
                )}
            </label>

            <h3>Jugador 2</h3>
            <label>
                Nombre:
                <input
                    type="text"
                    {...register('player2.name', {
                        required: 'Nombre es requerido',
                    })}
                    placeholder="Ingresa un nombre"
                />
                {errors.player2?.name && (
                    <span style={{ color: 'red' }}>
                        {errors.player2.name.message}
                    </span>
                )}
            </label>
            <label>
                Color:
                <select
                    {...register('player2.color', {
                        required: 'Selecciona un color',
                    })}
                >
                    {selectOptions.map((o) => (
                        <option key={o.value} value={o.value}>
                            {o.label}
                        </option>
                    ))}
                </select>
                {errors.player2?.color && (
                    <span style={{ color: 'red' }}>
                        {errors.player2.color.message}
                    </span>
                )}
            </label>

            <button>Iniciar Juego</button>
        </form>
    );
};
