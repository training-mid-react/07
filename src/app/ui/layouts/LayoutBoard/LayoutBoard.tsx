import './styles.scss';
import { LayoutBoardProps } from '@core/interfaces';

export const LayoutBoard = ({ children }: LayoutBoardProps) => {
    return <div className="board-layout">{children}</div>;
};
