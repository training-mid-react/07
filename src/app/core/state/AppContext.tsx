import { createContext, useContext, useReducer } from 'react';
import { initialState, reducer } from './reducer';
import { IContext, IState } from '@core/interfaces';

export const AppContext = createContext<IContext>({
    state: {} as IState,
    dispatch: () => {},
});

export const AppProvider = ({ children }: { children: React.ReactNode }) => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return (
        <AppContext.Provider value={{ state, dispatch }}>
            {children}
        </AppContext.Provider>
    );
};

export const useAppContext = () => useContext(AppContext);
