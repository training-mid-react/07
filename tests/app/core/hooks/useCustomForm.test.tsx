import { useCustomForm } from '@core/hooks';
import { RenderHookResult, renderHook, render } from '@testing-library/react';
import { vi, describe, test, expect, beforeEach } from 'vitest';
import { AppContext } from '../../../../src/app/core/state/AppContext';
import { LayoutBoard } from '../../../../src/app/ui/layouts/LayoutBoard/LayoutBoard';
import { initialState } from '../../../../src/app/core/state/reducer';
import { ICustomFormProps } from '../../../../src/app/core/interfaces/game/customFormProps.interface';
import { Form } from '../../../../src/app/ui/components/Form/index';

vi.mock('react-router-dom', () => {
    return {
        useNavigate: vi.fn(),
    };
});

let useCustomFormCall: RenderHookResult<ICustomFormProps, unknown>;

describe('useCustomForm hook', () => {
    const renderHookWithContext = (children: React.ReactNode) => {
        return render(
            <AppContext.Provider
                value={{ state: initialState, dispatch: vi.fn() }}
            >
                <LayoutBoard>{children}</LayoutBoard>
            </AppContext.Provider>
        );
    };

    beforeEach(() => {
        useCustomFormCall = renderHook(() => useCustomForm());
    });

    test('hooks respond correctly', () => {
        const { asFragment } = renderHookWithContext(
            <Form onSubmit={vi.fn()} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('should return correct select options', () => {
        const { result } = useCustomFormCall;

        expect(result.current.selectOptions).toEqual([
            { value: '', label: 'Selecciona un color' },
            { value: 'red', label: 'Rojo' },
            { value: 'blue', label: 'Azul' },
        ]);
    });
});
