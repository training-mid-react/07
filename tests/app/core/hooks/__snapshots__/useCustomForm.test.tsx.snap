// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useCustomForm hook > hooks respond correctly 1`] = `
<DocumentFragment>
  <div
    class="board-layout"
  >
    <form
      class="form"
    >
      <h2>
        Configuraci√≥n de Jugadores
      </h2>
      <h3>
        Jugador 1
      </h3>
      <label>
        Nombre:
        <input
          name="player1.name"
          placeholder="Ingresa un nombre"
          role="input"
          type="text"
        />
      </label>
      <label>
        Color:
        <select
          name="player1.color"
        >
          <option
            value=""
          >
            Selecciona un color
          </option>
          <option
            value="red"
          >
            Rojo
          </option>
          <option
            value="blue"
          >
            Azul
          </option>
        </select>
      </label>
      <h3>
        Jugador 2
      </h3>
      <label>
        Nombre:
        <input
          name="player2.name"
          placeholder="Ingresa un nombre"
          type="text"
        />
      </label>
      <label>
        Color:
        <select
          name="player2.color"
        >
          <option
            value=""
          >
            Selecciona un color
          </option>
          <option
            value="red"
          >
            Rojo
          </option>
          <option
            value="blue"
          >
            Azul
          </option>
        </select>
      </label>
      <button>
        Iniciar Juego
      </button>
    </form>
  </div>
</DocumentFragment>
`;
